README.md
# tailwindcss-var

This plugin allows to create and use scoped CSS variables for colors and spacings in your Tailwind CSS project.

For live-examples check out [mariohamann.de/tailwindcss-var](https://mariohamann.de/tailwindcss-var).

## Please read ❤️
- ⚠️ This plugin expects you to use Tailwind CSS >2.0 || >3.0 with JIT mode enabled.  (Tested only with >3.0)
- This is the first version of the plugin. Please be aware of changes in future.
- The plugin works best with the default color variations of Tailwind CSS (50, 100, ..., 800, 900). If you want to use other variable variations, you'll find some basic instructions below.
## Installation

1. Add package to your project e. g. `npm install @mariohamann/tailwindcss-var`.
2. Add the plugin to your `tailwind.config.js` with `require('@mariohamann/tailwindcss-var')`.
3. Add the following color variables to your `tailwind.config.js` (or change them to your needs as described [below](#modify-color-variables)):

```js
{
  theme: {
    extend: {
      colors: {
        var: {
          50: 'var(--tw-var-color-50)',
          100: 'var(--tw-var-color-100)',
          200: 'var(--tw-var-color-200)',
          300: 'var(--tw-var-color-300)',
          400: 'var(--tw-var-color-400)',
          500: 'var(--tw-var-color-500)',
          600: 'var(--tw-var-color-600)',
          700: 'var(--tw-var-color-700)',
          800: 'var(--tw-var-color-800)',
          900: 'var(--tw-var-color-900)',
        },
      },
    },
  }
}
```

4. To test if everything works, add the following classes to an object with text `var-indigo text-var-500`.
## Examples
### Example 1: Basic color change

1. Define your object once with text-var-600, bg-var-50 etc.
2. Fill the variable (dynamically) with e. g. var-pink.
3. text-var-600 renders as text-pink-600, bg-var-50 renders as bg-pink-50.

Try it yourself: https://play.tailwindcss.com/8IoJCK0r1t.

![tailwindcss-var netlify app_ (8)](https://user-images.githubusercontent.com/26542182/152775017-8ef5048b-190b-4568-9b8a-1bbdeed71a8b.png)


```html
<button class="var-... text-var-600 bg-var-50 border-var-600 hover:bg-var-100 focus:ring-var-500 ...">...</button>
```

### Example 2: Hierarchy and subvalues

- Variables are automatically consumable by every child element.
- You can overwrite variables in every child.
- You can overwrite single variants of the variable with other colors or opacities.

Try it yourself: https://play.tailwindcss.com/fQPslGpLCw

![tailwindcss-var netlify app_ (7)](https://user-images.githubusercontent.com/26542182/152775079-41abeb53-c3f5-4cae-a187-ca14bf189695.png)


```html
<div class="var-...">
  <button class="text-white bg-var-600 border-var-800 hover:bg-var-700 focus:ring-var-500 ...">...</button>
  <button class="text-var-600 bg-var-50 border-var-600 hover:bg-var-100 focus:ring-var-500 ...">...</button>
  <button class="text-var-600 bg-var-50 border-var-50 hover:bg-var-100 focus:ring-var-500 ...">...</button>
  <button class="var-600-.../20 text-var-600 bg-var-50 border-var-50 ..." Disabled>...</button>
</div>
```

### Example 3: Sizes

- Besides color variables you can set variables for spacings.
- You can use the variables for width, height, margins etc.

Try it yourself: https://play.tailwindcss.com/ojWbNZGdMY

![tailwindcss-var netlify app_ (5)](https://user-images.githubusercontent.com/26542182/152775109-7f923575-36b6-4d88-91ee-7e242ec0e732.png)


```html
<div class="var-spacing-... w-var h-var ...">...</button>
```

### Example 4: Arbitrary values
- You can set and use arbitrary values for both spacings and colors.
- In the example height and roundedness resize perfectly with the width of the object.

Try it yourself: https://play.tailwindcss.com/ETCMSUlWiv

![tailwindcss-var netlify app_ (6)](https://user-images.githubusercontent.com/26542182/152775129-7e00e1e4-cb96-485d-bc28-40c6001bde6b.png)


## Modify color variables
If you use Tailwind's default color definitions, your theme should be extended with the following (as described above):

```js
{
  theme: {
    extend: {
      colors: {
        var: {
          50: 'var(--tw-var-color-50)',
          100: 'var(--tw-var-color-100)',
          200: 'var(--tw-var-color-200)',
          300: 'var(--tw-var-color-300)',
          400: 'var(--tw-var-color-400)',
          500: 'var(--tw-var-color-500)',
          600: 'var(--tw-var-color-600)',
          700: 'var(--tw-var-color-700)',
          800: 'var(--tw-var-color-800)',
          900: 'var(--tw-var-color-900)',
        },
      },
    },
  }
}
```

If you have other color definitions which you want to use with variables, just extend the `var`-definitions in your theme e. g.: 

```js
{
  theme: {
    extend: {
      colors: {
        primary: {
          light: '#FF88DD',
          dark: '#FFFF00',
        }
        secondary: {
          light: 'FF88FF',
          dark: '#FF00FF',
        }
        var: {
          light: 'var(--tw-var-color-light)',
          dark: 'var(--tw-var-color-dark)',
        },
      },
    },
  }
}
```

In the future the plugin will set overwritable default color variables.


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [0.1.0] - 2022-02-06

Initial release

[0.1.0]: https://github.com/tailwindlabs/tailwindcss-line-clamp/releases/tag/v0.1.0


src/index.js
const plugin = require('tailwindcss/plugin')

const tailwindcssVar = plugin(
  function ({ matchUtilities, theme }) {
    if (theme('colors.var', {})) {
      const variablesPalette = Object.keys(theme('colors.var', {}))

      /*
       * Set one-level variables like var-red, var-indigo, var-[pink] etc.
       */

      matchUtilities(
        {
          var: (color) => {
            if (typeof color === 'string') {
              const output = {}
              variablesPalette.map((variant) => {
                output[`--tw-var-color-${variant}`] = color
              })
              return output
            } else {
              const output = {}
              variablesPalette.map((variant) => {
                if (color[variant]) {
                  output[`--tw-var-color-${variant}`] = color[variant]
                } else if (color['DEFAULT']) {
                  output[`--tw-var-color-${variant}`] = color['DEFAULT']
                }
              })
              return output
            }
          },
        },
        { values: theme('colors', {}), type: 'color' }
      )

      /*
       * Set two-level variables like var-50-red, var-100-indigo, var-50-[pink] etc.
       * We need to transform the color palette from format "color.shade" to "shade.color"
       * Examples:
       * - red.50 ... red.900  ->  50.red   ... 900.red
       * - black               ->  50.black ... 900.black
       */

      const colorsAsEntries = Object.entries(theme('colors', {}))

      const getTransformedColors = (variant) => {
        const transformedColors = {}
        colorsAsEntries.map(([key, value]) => {
          if (typeof value === 'string') {
            transformedColors[key] = value
          } else {
            if (value[variant]) {
              transformedColors[key] = value[variant]
            } else if (value['DEFAULT']) {
              transformedColors[key] = value['DEFAULT']
            }
          }
        })
        return transformedColors
      }

      variablesPalette.map((variant) => {
        matchUtilities(
          {
            [`var-${variant}`]: (value) => ({
              [`--tw-var-color-${variant}`]: value,
            }),
          },
          { values: getTransformedColors(variant), type: 'color' }
        )
      })
    }

    /*
     * Set spacing variables like var-spacing-2 etc. to be used with h-var, w-var, m-var etc.
     *
     * It would have been nicer to just use "var-2" instead of "var-spacing-2" to set the variable.
     * Unfortunately this would collide with color variables when setting arbitrary values.
     *
     * Furthermore it would have been nicer to use "size" instead of "spacing",
     * but I want to stick to the official Tailwind terminology.
     */

    if (theme('spacing.var', {})) {
      matchUtilities(
        {
          'var-spacing': (spacing) => ({
            '--tw-var-spacing': spacing,
          }),
        },
        { values: theme('spacing', {}) }
      )
    }
  },
  {
    theme: {
      extend: {
        spacing: {
          var: 'var(--tw-var-spacing)',
        },
      },
    },
  }
)

module.exports = tailwindcssVar
